pkg/clustermesh/mcsapi/serviceimport_controller.go

```diff
@@ -0,0 +1,542 @@
+		mcsAPISvcType = mcsapiv1alpha1.Headless
+	}
+	mcsAPISvcSpec := &mcsapitypes.MCSAPIServiceSpec{
+		Name:                  svc.Name,
+		Namespace:             svc.Namespace,
+		Ports:                 ports,
+		Type:                  mcsAPISvcType,
+		SessionAffinity:       svc.Spec.SessionAffinity,
+		SessionAffinityConfig: svc.Spec.SessionAffinityConfig.DeepCopy(),
+	}
+	mcsAPISvcSpec.ExportCreationTimestamp = svcExport.CreationTimestamp
+	return mcsAPISvcSpec
+}
+
+func getOldestRemoteClusterService(svcExportByCluster operator.ServiceExportsByCluster) string {
+	oldestCluster := maps.Keys(svcExportByCluster)[0]
+	for cluster, currentSvc := range svcExportByCluster {
+		if currentSvc.ExportCreationTimestamp.Before(
+			&svcExportByCluster[oldestCluster].ExportCreationTimestamp) {
+			oldestCluster = cluster
+		}
+	}
+	return oldestCluster
+}
+
+type portMergeKey struct {
+	Protocol corev1.Protocol
+	Port     int32
+}
+
+type portMergeValue struct {
+	mcsapiv1alpha1.ServicePort
+
+	clusterName             string
+	ExportCreationTimestamp metav1.Time
+}
+
+type mergedPorts map[portMergeKey]portMergeValue
+
+// checkConflictDuplicatedPortName checks if two ports with the same name are similar
+// (same port and proto). If they are the same and one of the port is from the
+// local cluster this function returns a conflict. Also the port key is returned for
+// the port that have lesser precedence if both ports conflict.
+func checkConflictDuplicatedPortName(existingConflict, localClusterName string, val1, val2 portMergeValue) (portMergeValue, string, *portMergeKey) {
+	key1 := portMergeKey{
+		Protocol: val1.Protocol,
+		Port:     val1.Port,
+	}
+	key2 := portMergeKey{
+		Protocol: val2.Protocol,
+		Port:     val2.Port,
+	}
+
+	currVal := val1
+	deleteKey := &key2
+	if val2.ExportCreationTimestamp.Before(&val1.ExportCreationTimestamp) {
+		currVal = val2
+		deleteKey = &key1
+	}
+	if key1 == key2 {
+		deleteKey = nil
+	}
+
+	hasLocalCluster := val1.clusterName == localClusterName || val2.clusterName == localClusterName
+	if existingConflict != "" || !hasLocalCluster || deleteKey == nil {
+		return currVal, existingConflict, deleteKey
+	}
+
+	return currVal, fmt.Sprintf(
+		"Duplicated port name \"%s\". Using port definition \"%v\" from oldest service export in cluster \"%s\".",
+		currVal.Name, currVal, currVal.clusterName,
+	), deleteKey
+}
+
+// checkConflictDuplicatedPort checks if two similar ports (same port and proto)
+// have the same name.
+func checkConflictDuplicatedPort(existingConflict, localClusterName string, val1, val2 portMergeValue) (portMergeValue, string) {
+	currVal := val1
+	if val2.ExportCreationTimestamp.Before(&val1.ExportCreationTimestamp) {
+		currVal = val2
+	}
+	hasLocalCluster := val1.clusterName == localClusterName || val2.clusterName == localClusterName
+	if existingConflict != "" || !hasLocalCluster || (val1.Name == val2.Name && ptr.Deref(val1.AppProtocol, "") == ptr.Deref(val2.AppProtocol, "")) {
```

Comment:
The `hasLocalCluster` was more meant to not raise conflict messages between two remote clusters (aka the local cluster is not present in this conflict) so like we do not raise conflict in this scenario for instance:
`local` -> no conflict with `oldest`
`oldest` -> conflict with `other`
`other` -> conflict with `oldest`

Reply:
Hmm, my consideration was based on the following wording from the KEP:

> Conflict resolution policy: If any properties have conflicting values that can not simply be merged, a ServiceExportConflict condition will be set on all ServiceExports for the conflicted service with a description of the conflict. The conflict will be resolved by assigning precedence based on each ServiceExport's creationTimestamp, from oldest to newest.

which suggests that the conflict condition should be set to all ServiceExports associated with that service. However, it is not totally clear whether that applies to ports conflict resolution as well, given that they are a bit special compared to the other properties.

Reply:
Arf indeed I will try to ask around because I am not sure this is super logical. Like for instance with the type of conflict message suggested in the KEP there is no way you would know that there is actually no problem with your local cluster as it only suggest the winning cluster (like in the example I gave above you would have a conflict message indicating that `oldest` win but no that the conflict was between `oldest` and `other` which is IMO confusing if you are looking it from the `local` cluster view).

Reply:
So I asked in #sig-multicluster on the kubernetes slack (https://kubernetes.slack.com/archives/C09R1PJR3/p1725449448075499) and Tom Pantelis, a Submariner maintainer, answered that they were not technically able to respect that on the Submariner project (apparently they don't have all the service exports info from all the clusters like we do) and that he would like to loosen the language on the KEP so that their implementation would be conformant. 

I don't think we would have more answers until the next sig-multicluster meeting which should be happening the 18/09.  I will try to ask again the same question there to see what should ideally be happening from my little 3 clusters example mentioned in this thread and if as a result we should update the KEP or not.